caddy: add suport for compiling Caddy with plugins

vincentbernat opened the pull request:
This adds a `withPlugins` function to Caddy package.

```nix
services.caddy = {
  enable = true;
  package = pkgs.caddy.withPlugins {
    plugins = [ "github.com/caddy-dns/powerdns@v1.0.1" ];
    hash = "sha256-F/jqR4iEsklJFycTjSaW8B/V3iTGqqGOzwYBUXxRKrc=";
  };
};
```

Fix: #14671 

This is an alternative to #317881 and it relies on xcaddy. Looking at https://github.com/NixOS/nixpkgs/pull/317881#issuecomment-2258807399, I am still missing tests. I am unsure if this is a build test (in `checkPhase`) or a NixOS test. The remaining requirements should be OK (notably use of xcaddy and FOD).

The release notes are missing, I'll add them once this gets a chance to be accepted.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

stepbrobd commented on pkgs/by-name/ca/caddy/package.nix:
```diff
@@ -61,12 +65,53 @@ buildGoModule {
+      };
     };
+    withPlugins =
+      { plugins
```
Quick question, should we make the build fail if a plugin is provided without the version string?

For example:

```nix
pkgs.caddy.withPlugins {
  plugins = [ "github.com/caddy-dns/powerdns@v1.0.1" "github.com/caddy-dns/cloudflare" ];
  hash = "sha256-AoW35l7QkXunjBzZ43IlyU3UkVXw2D4eyc1jx8xpT0U=";
}
```

After testing this on my darwin machine, I have the following:


```console
$ /nix/store/icp2z20hpf2ps7g4n5rzqdkg5qsjp38z-caddy-2.8.4/bin/caddy build-info
...
dep	github.com/caddy-dns/cloudflare	v0.0.0-20240703190432-89f16b99c18e
dep	github.com/caddy-dns/powerdns	v1.0.1	
...
```

vincentbernat replied:
Yes, I think I can add an assertion.

vincentbernat replied:
I have added an assertion (using the first non-versioned plugin as an example).

stepbrobd commented on pkgs/by-name/ca/caddy/package.nix:
```diff
@@ -61,12 +65,53 @@ buildGoModule {
+          pluginsList = lib.concatMapStrings (plugin: "${plugin}-") pluginsSorted;
+          pluginsHash = builtins.hashString "sha1" pluginsList;
+        in stdenv.mkDerivation {
+          pname = "caddy-src-with-plugins-${pluginsHash}";
```
Small nit, imo `pluginsHash` makes the build log a bit too long, maybe `caddy-src-with-plugins` is good enough here?

vincentbernat replied:
It's there to ensure a cached build is not used when adding/removing a plugin. This was one of the request in https://github.com/NixOS/nixpkgs/pull/317881#issuecomment-2258807399.

vincentbernat replied:
I have switched to md5 to reduce the length a bit. We could also use a subset of the hash if it's still too long.

stepbrobd replied:
I'm not sure if using md5 is a good idea, IIRC it's considered outdated and insecure (citation needed)?

I took a look into the comment you linked, it seems like it'd be a good idea to include a test to make sure specified plugins are properly installed. I came up with the following but don't have time to dig deeper (feel free to take whatever you need):

```diff
diff --git a/pkgs/by-name/ca/caddy/package.nix b/pkgs/by-name/ca/caddy/package.nix
index eea6894ce328..c052da5ef290 100644
--- a/pkgs/by-name/ca/caddy/package.nix
+++ b/pkgs/by-name/ca/caddy/package.nix
@@ -116,6 +116,31 @@ buildGoModule {
             outputHash = hash;
             outputHashAlgo = "sha256";
           };
+
+          doInstallCheck = true;
+          installCheckPhase = ''
+            runHook preInstallCheck
+
+            build_info="$($out/bin/caddy build-info)"
+
+            for plugin in ''${plugins[@]}; do
+              # this won't work :(
+              echo $plugin
+              url=$(echo "$plugin" | cut -d'@' -f1)
+              version=$(echo "$plugin" | cut -d'@' -f2)
+              echo $url
+              echo $version
+
+              if echo "$build_info" | grep -q "$url[[:space:]]*$version"; then
+                echo "$plugin found in build-info"
+              else
+                echo "$plugin not found in build-info" >&2
+                exit 1
+              fi
+            done
+
+            runHook postInstallCheck
+          '';
       });
   };

```

For testing:

```bash
nom-build --expr 'with import ./. { }; caddy.withPlugins { plugins = [ "github.com/caddy-dns/powerdns@v1.0.1" "github.com/caddy-dns/cloudflare@v0.0.0-20240703190432-89f16b99c18e" ]; hash = "sha256-AoW35l7QkXunjBzZ43IlyU3UkVXw2D4eyc1jx8xpT0U="; }'
```

vincentbernat replied:
I have added a postInstallCheck as you suggest.

As for the MD5, this is not meant to be secure, just a safety to ensure the user does not forget to update the hash when modifying plugins. A mechanism like this was requested in the original PR.

stepbrobd commented:
LGTM!

Tested with couple more plugins and everything works fine on aarch64-darwin and x86_64-linux

stepbrobd commented:
cc @NickCao

NickCao commented on pkgs/by-name/ca/caddy/package.nix:
```diff
@@ -61,12 +65,82 @@ buildGoModule {
+            xcaddy
+            cacert
+          ];
+          unpackPhase = "true";
```
```suggestion
          dontUnpack = true;
```

vincentbernat replied:
Fixed. Thanks.

NickCao reviewed (commented):
Apart from the nitpick, looks pretty good!

BatteredBunny commented:
fyi theres a typo in the commit message

vincentbernat commented:
I have also updated the release notes.

Bear-03 commented:
I am trying to use this with the `caddy-tailscale` plugin and I am getting the following error:

```
error: builder for '/nix/store/jqhihzbbylzykyr6rv0bw1zmx0bs2csi-caddy-src-with-xcaddy-2.8.4.drv' failed with exit code 1;
       last 25 log lines:
       > go: added github.com/prometheus/client_golang v1.19.1
       > go: added github.com/prometheus/client_model v0.5.0
       > go: added github.com/prometheus/common v0.48.0
       > go: added github.com/prometheus/procfs v0.12.0
       > go: added github.com/quic-go/qpack v0.4.0
       > go: added github.com/quic-go/quic-go v0.44.0
       > go: added github.com/zeebo/blake3 v0.2.3
       > go: added go.uber.org/mock v0.4.0
       > go: added go.uber.org/multierr v1.11.0
       > go: added go.uber.org/zap v1.27.0
       > go: added go.uber.org/zap/exp v0.2.0
       > go: added golang.org/x/crypto v0.23.0
       > go: added golang.org/x/exp v0.0.0-20240506185415-9bf2ced13842
       > go: added golang.org/x/mod v0.17.0
       > go: added golang.org/x/net v0.25.0
       > go: added golang.org/x/sync v0.7.0
       > go: added golang.org/x/sys v0.20.0
       > go: added golang.org/x/term v0.20.0
       > go: added golang.org/x/text v0.15.0
       > go: added golang.org/x/time v0.5.0
       > go: added golang.org/x/tools v0.21.0
       > go: added google.golang.org/protobuf v1.34.1
       > 2024/12/03 10:48:11 [INFO] exec (timeout=0s): /nix/store/bavnchxi7v6xs077jxv7fl5rrqc3y87w-go-1.23.3/bin/go get -v 
github.com/tailscale/caddy-tailscale/v2@v2.8.4 github.com/caddyserver/caddy/v2@v2.8.4
       > go: github.com/tailscale/caddy-tailscale/v2@v2.8.4: git init --bare in 
/build/go/pkg/mod/cache/vcs/b8e7a75999f9e1392c6e6d48305440fe816b1297c9aeca618521302caf9073f0: exec: "git": executable file not found in $PATH
       > 2024/12/03 10:48:12 [FATAL] exit status 1
       For full logs, run 'nix log /nix/store/jqhihzbbylzykyr6rv0bw1zmx0bs2csi-caddy-src-with-xcaddy-2.8.4.drv'.
error: 1 dependencies of derivation '/nix/store/8hbr7zb91mgbi7nv8fzgarhv358s0ag0-caddy-2.8.4.drv' failed to build
error: 1 dependencies of derivation '/nix/store/hj2f4hp41v2b394381mmlv5s5g0r7rib-Caddyfile-formatted.drv' failed to build
error: 1 dependencies of derivation '/nix/store/flcr0znjba8f789rkwnhs7j38hqqzx1l-etc.drv' failed to build
error: 1 dependencies of derivation '/nix/store/4d1xka9akhkg5lpw639hbmgbqi1mmc0j-nixos-system-nixos-laptop-24.11.20241130.62c435d.drv' failed to build
[11:48:13] ERROR    Command 'sudo nixos-rebuild switch --flake /home/bear/.dotfiles#nixos-laptop' failed with exit code 1                    console.py:183
           ERROR    System update failed: Command '['sudo', 'nixos-rebuild', 'switch', '--flake', '/home/bear/.dotfiles#nixos-laptop']' returned  app.py:61
                    non-zero exit status 1.
```

This is my code:

```nix
services.caddy = {
  enable = true;
  package = pkgs.caddy.withPlugins {
    plugins = [ "github.com/tailscale/caddy-tailscale@v2.8.4" ];
    hash = ""; # It doesn't build at all so i haven't been able to get the hash
  };
};
```

vincentbernat commented:
I have added `git` as a native build input, as it may be required for modules not present on `proxy.golang.org`. There was a comment about that. `goModule` also uses git as a native build input.

kusold commented on pkgs/by-name/ca/caddy/package.nix:
```diff
@@ -61,12 +66,83 @@ buildGoModule {
+            done
+          done < <($out/bin/caddy build-info)
+          if (( ''${#notfound[@]} )); then
+            >&2 echo "Plugins not found: ''${notfound[@]}"
```
I ran into an issue with this check when plugin versions are designed as git-shas.

When given the input:
```
    plugins = [
      "github.com/caddy-dns/cloudflare@89f16b99c18ef49c8bb470a82f895bce01cbaece"
      "github.com/dulli/caddy-wol@c0d58507c9037191aa9622d531a001db619dd543"
    ];
    hash = "sha256-pKiL3bGcFVcD7r67mY9RtAEHDJc+gq3vz8DkZYssfb4=";
```

caddy build-info contains:
```
...
dep     github.com/caddy-dns/cloudflare v0.0.0-20240703190432-89f16b99c18e
...
dep     github.com/dulli/caddy-wol      v1.0.1-0.20240903185854-c0d58507c903
...
```
This causes this check to fail because the versions were substituted. If I update my input to the versions, it works:

```
plugins = [
      "github.com/caddy-dns/cloudflare@v0.0.0-20240703190432-89f16b99c18e"
      "github.com/dulli/caddy-wol@v1.0.1-0.20240903185854-c0d58507c903"
    ];
    hash = "sha256-pKiL3bGcFVcD7r67mY9RtAEHDJc+gq3vz8DkZYssfb4=";
```


Do you need to check for the version, or is matching the module sufficient?

stepbrobd replied:
IMO checking module + version is better, and you can get the format it wants beforehand with `go get` and check the `go.mod` file

thegatesdev commented:
Is there intent to support plugins like [FrankenPHP](https://frankenphp.dev/docs/compile/)?
FrankenPHP requires PHP to be present at compile- and runtime and uses CGO.

vincentbernat commented:
I would need to test, but I think this is easily doable. However, maybe in a second PR once this one is merged?

stepbrobd commented:
@vincentbernat there's a merge conflict

btw I think you can ping `NickCao` for merging after the conflict's resolved

vincentbernat committed:
caddy: add support for compiling Caddy with plugins

This adds a `withPlugins` function to Caddy package.

```nix
services.caddy = {
  enable = true;
  package = pkgs.caddy.withPlugins {
    plugins = [ "github.com/caddy-dns/powerdns@v1.0.1" ];
    hash = "sha256-F/jqR4iEsklJFycTjSaW8B/V3iTGqqGOzwYBUXxRKrc=";
  };
};
```

vincentbernat commented:
Rebased.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/2153

thiagokokada commented on pkgs/by-name/ca/caddy/package.nix:
```diff
@@ -61,12 +66,83 @@ buildGoModule {
+        package = caddy;
+      };
     };
+    withPlugins =
```
Nit, but I think this is very long function to be defined at the body of the derivation. I suggest moving this to `with-plugins.nix` or something and importing the file here.

thiagokokada replied:
I will not block PR for my nit, so if someone else wants to merge feel free.

vincentbernat replied:
Done, moved to `plugins.nix` and use of `callPackage`. I think this is the right way, but maybe I should inherit more stuff?

thiagokokada commented on pkgs/by-name/ca/caddy/package.nix:
```diff
@@ -61,12 +66,83 @@ buildGoModule {
+      { plugins
+      , hash ? lib.fakeHash
+      }: caddy.overrideAttrs (finalAttrs: prevAttrs:
+      let
+        pluginsSorted = builtins.sort builtins.lessThan plugins;
+        pluginsList = lib.concatMapStrings (plugin: "${plugin}-") pluginsSorted;
+        pluginsHash = builtins.hashString "md5" pluginsList;
+        pluginsWithoutVersion = builtins.filter (p: !lib.hasInfix "@" p) pluginsSorted;
+      in
+      assert lib.assertMsg (builtins.length pluginsWithoutVersion == 0)
```
Another non-blocking nit, but in general `builtins` should be avoided because sometimes we have polyfills for older versions of Nix inside `lib`, so in general we should always prefer using `lib`.

vincentbernat replied:
Done, except for `builtins.hashString` as there does not seem there is an equivalent.

vincentbernat committed:
caddy: move withPlugins function into a dedicated file

vincentbernat committed:
caddy: avoid use of builtins and prefer lib

There is no `lib.hashString`, so keep using `builtins` for this one.

oxalica commented:
It does not work well with a commit-hash version. I'm using https://github.com/mholt/caddy-webdav which does not have a version tag, and it results in an unhelpful build error.

```nix
caddy.withPlugins {
  plugins = [
    "github.com/mholt/caddy-webdav@42168ba04c9dc2cd228ab8c453dbab27654e52e6"
  ];
  hash = "";
}
```

```
caddy> Running phase: installCheckPhase
caddy> Plugins not found: github.com/mholt/caddy-webdav@42168ba04c9dc2cd228ab8c453dbab27654e52e6
```

Thanks to @kusold 's https://github.com/NixOS/nixpkgs/pull/358586#discussion_r1868783011, I need to fill in the magic `v0.0.0-{date}-{shortrev}` to make it build. (Sorry if it's in any obvious format but I'm not a Go dev)

If this is intentional, could we improve the documentation to inform how to generate a valid version string for commit-hash-only plugins? Or if it's possible, could we provide these information directly in build error messages?

stepbrobd commented:
What I usually do is create an empty project and run `go get` on the plugin to get the full string:

```console
$ go mod init temp
$ go get github.com/mholt/caddy-webdav
$ grep 'caddy-webdav' go.mod
        github.com/mholt/caddy-webdav v0.0.0-20241008162340-42168ba04c9d // indirect

oxalica commented:
Is it possible to override the caddy source when using `withPlugin`? It seems `(caddy.overrideAttrs (old: { patches = [ .. ]; })).withPlugin { .. }` will not applying the patch at all. Overriding `src` or `version` to use a different version also has no effect.

I tried `(caddy.withPlugin { .. }).overrideAttrs (old: { patches = [ .. ]; })`. It does try to apply the patch, but fails because the source layout is significantly different than upstream.

stepbrobd commented:
Oof this is a hard one

`withPlugin` overrides the caddy source by using the one generated by xcaddy, i.e. if you call `withPlugins`, the src will be replaced (from the full caddy source) to only one `main.go` and vendored dependencies (where the full caddy source will be moved to)

If you want to apply the patches, try changing the patch paths to `/vendor/github.com/caddyserver/caddy/v2/...`. In other words, instead of apply patches directly on `(caddy.withPlugins {...}).src`, apply them to `"${(caddy.withPlugins {...}).src}/vendor/github.com/caddyserver/caddy/v2/"`

To illustrate:

```console
$ # im in nixpkgs
$ nix repl -f .
Type :? for help.
Loading installable ''...
Added 23515 variables.
nix-repl> :b (caddy.withPlugins {
          plugins = [
                "github.com/caddy-dns/cloudflare@v0.0.0-20240703190432-89f16b99c18e"
          ];
          hash = "sha256-JoujVXRXjKUam1Ej3/zKVvF0nX97dUizmISjy3M3Kr8=";
          }).src

This derivation produced the following outputs:
  out -> /nix/store/48vdfpdaz04wcdqvymx7cx29vsav5pxj-caddy-src-with-plugins-0bbf432f07a600413ee833233180f51d-2.9.0
nix-repl> :b caddy.src

This derivation produced the following outputs:
  out -> /nix/store/vwmnpjxxfq1ffc5a82c0q24p0c3sdl3h-source
$ # withPlugins caddy src:
$ > ls -l /nix/store/48vdfpdaz04wcdqvymx7cx29vsav5pxj-caddy-src-with-plugins-0bbf432f07a600413ee833233180f51d-2.9.0/vendor/github.com/caddyserver/caddy/v2
total 608
-r--r--r--   1 root  nixbld    399 Dec 31  1969 AUTHORS
-r--r--r--   1 root  nixbld  11358 Dec 31  1969 LICENSE
-r--r--r--   1 root  nixbld  12002 Dec 31  1969 README.md
-r--r--r--   1 root  nixbld  45621 Dec 31  1969 admin.go
-r--r--r--   1 root  nixbld  33278 Dec 31  1969 caddy.go
...
$ # caddy.src:
$ ls -l /nix/store/vwmnpjxxfq1ffc5a82c0q24p0c3sdl3h-source
total 880
-r--r--r--   1 root  nixbld    399 Dec 31  1969 AUTHORS
-r--r--r--   1 root  nixbld  11358 Dec 31  1969 LICENSE
-r--r--r--   1 root  nixbld  12002 Dec 31  1969 README.md
-r--r--r--   1 root  nixbld  45621 Dec 31  1969 admin.go
-r--r--r--   1 root  nixbld   5790 Dec 31  1969 admin_test.go
-r--r--r--   1 root  nixbld  33278 Dec 31  1969 caddy.go
...
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tailscale-caddy-certificates-not-working/61228/6

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/using-nix-to-replace-xcaddy-to-build-caddy-webserver/11765/3

Jonas-Sander commented:
Please add this to the docs!     
I wasted hours, because I didn't know about this way.

aaravrav commented:
How can I replace a caddy module with my own? This is implemented in [xcaddy's --with flag](https://github.com/caddyserver/xcaddy?tab=readme-ov-file#custom-builds)

sokai commented:
@stepbrobd´: Thanks a lot for your/this implementation + dedication!
May I ask you, if/how it's possible to use the `GOPROXY` env var during the building process? – I tried a lot, with no luck, to use my [Athens](https://github.com/NixOS/nixpkgs/blob/release-25.05/nixos/modules/services/development/athens.md) proxy, that works with `go install $MODULE$` but not with `nixos-rebuild`).
KR

sokai commented:
Thanks for your fast reply! :)

> using `overrideAttrs` to inject `GOPROXY` into the build env var

Yes, think so …

**Current/simple setup**
```
 caddyWithPlugins = pkgs.caddy.withPlugins {
    plugins = [
      […]
    ];
    hash = "sha256-9TR2vFgrxupy8O4i+a05+/Fj2D4sQCw42q2odX22fD0=";
  };

```

**`overrideAttrs` setup**
```
      caddyWithPlugins = (pkgs.caddy.overrideAttrs (
        finalAttrs: prevAttrs: {
          env.GOPROXY = "http://$MYATHENSPROXYHOSTNAME:3000,https://proxy.golang.org,direct";
        }
      )).withPlugins {
    plugins = [
      […]
    ];
    hash = "sha256-9TR2vFgrxupy8O4i+a05+/Fj2D4sQCw42q2odX22fD0=";
  };
```

With both setups `$GOPROXY` isn't used … :/
